cachito/web/migrations/versions/1714d8e3002b_map_deps_to_pkg.py

```diff
@@ -0,0 +1,136 @@
+
+from alembic import op
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = "1714d8e3002b"
+down_revision = "b46cf36806d7"
+branch_labels = None
+depends_on = None
+
+log = logging.getLogger("alembic")
+
+request_dependency_table = sa.Table(
+    "request_dependency",
+    sa.MetaData(),
+    sa.Column("request_id", sa.Integer()),
+    sa.Column("dependency_id", sa.Integer()),
+    sa.Column("package_id", sa.Integer()),
+)
+
+request_package_table = sa.Table(
+    "request_package",
+    sa.MetaData(),
+    sa.Column("request_id", sa.Integer()),
+    sa.Column("package_id", sa.Integer()),
+)
+
+package_table = sa.Table(
+    "package",
+    sa.MetaData(),
+    sa.Column("id", sa.Integer(), primary_key=True),
+    sa.Column("name", sa.String()),
+    sa.Column("type", sa.String()),
+)
+
+
+def upgrade():
+    with op.batch_alter_table("request_dependency") as batch_op:
+        # Temporarily make this column nullable so it can be populated in the data migration
+        batch_op.add_column(
+            sa.Column("package_id", sa.Integer(), autoincrement=False, nullable=True)
+        )
+        batch_op.create_foreign_key(
+            "fk_request_dependency_package_id", "package", ["package_id"], ["id"]
+        )
+        batch_op.drop_constraint("request_dependency_request_id_dependency_id_key")
+
+    _upgrade_data()
+
+    with op.batch_alter_table("request_dependency") as batch_op:
+        batch_op.alter_column("package_id", existing_type=sa.INTEGER(), nullable=False)
+        batch_op.create_index(
+            batch_op.f("ix_request_dependency_package_id"), ["package_id"], unique=False
+        )
+        batch_op.create_unique_constraint(
+            "request_dependency_request_id_dependency_id_package_id_key",
+            ["request_id", "dependency_id", "package_id"],
+        )
+
+
+def _upgrade_data():
+    connection = op.get_bind()
+    last_request_id = None
+    for request_dep in connection.execute(
+        request_dependency_table.select().order_by(request_dependency_table.c.request_id)
+    ).fetchall():
+        if last_request_id != request_dep.request_id:
+            last_request_id = request_dep.request_id
+            log.info(
+                "Associating packages with the dependencies for request %d", request_dep.request_id
+            )
+
+        # Note that a package can be a top-level package or dependency
+        dependency = connection.execute(
+            package_table.select().where(package_table.c.id == request_dep.dependency_id)
+        ).fetchone()
+        package = connection.execute(
+            package_table.select()
+            .select_from(
+                package_table.join(
+                    request_package_table, package_table.c.id == request_package_table.c.package_id
+                )
+            )
+            .where(request_package_table.c.request_id == request_dep.request_id)
+            .where(package_table.c.type == dependency.type)
```

Comment:
Why is dependency type needed if we're fetching by ID?

Reply:
@lcarva a request could have multiple package managers associated with it, so we want to get the package that is of the same type as the dependency.